File: src/main/java/vjson/util/Transformer.java
/*
 * The MIT License
 *
 * Copyright 2019 wkgcass (https://github.com/wkgcass)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package vjson.util;

import vjson.JSON;
import vjson.simple.*;

import java.util.*;
import java.util.function.Function;

@SuppressWarnings("WeakerAccess")
public class Transformer {
    private final Map<Class<?>, Function<Object, JSON.Instance>> rules = new LinkedHashMap<>();

    public Transformer() {
        this.
            addRule(Boolean.class, SimpleBool::new)
            .addRule(Integer.class, SimpleInteger::new)
            .addRule(Long.class, SimpleLong::new)
            .addRule(Double.class, SimpleDouble::new)
            .addRule(Float.class, (Function<Float, JSON.Instance>) SimpleDouble::new)
            .addRule(Number.class, x -> new SimpleInteger(x.intValue()))
            .addRule(String.class, SimpleString::new)
            .addRule(Character.class, c -> new SimpleString(Character.toString(c)));
    }

    public <T> Transformer addRule(Class<T> type, Function<T, JSON.Instance> func) {
        //noinspection unchecked
        rules.put(type, (Function<Object, JSON.Instance>) func);
        return this;
    }

    public Transformer removeRule(Class type) {
        rules.remove(type);
        return this;
    }

    public JSON.Instance transform(Object input) {
        if (input == null)
            return new SimpleNull();
        if (input instanceof JSON.Instance) {
            return (JSON.Instance) input;
        }
        if (input instanceof Collection) {
            Collection coll = (Collection) input;
            List<JSON.Instance> list = new ArrayList<>(coll.size());
            for (Object o : coll) {
                list.add(transform(o));
            }
            return new SimpleArray(list, TrustedFlag.FLAG) {
            };
        }
        if (input instanceof Map) {
            Map inputMap = (Map) input;
            List<SimpleObjectEntry<JSON.Instance>> map = new ArrayList<>(inputMap.size());
            for (Object key : inputMap.keySet()) {
                if (!(key instanceof String)) {
                    throw new IllegalArgumentException("keys of map should be String");
                }
                map.add(new SimpleObjectEntry<>((String) key, transform(inputMap.get(key))));
            }
            return new SimpleObject(map, TrustedFlag.FLAG) {
            };
        }
        Class<?> type = input.getClass();
        for (Map.Entry<Class<?>, Function<Object, JSON.Instance>> entry : rules.entrySet()) {
            if (entry.getKey().isAssignableFrom(type)) {
                return entry.getValue().apply(input);
            }
        }
        throw new IllegalArgumentException("unknown input type: " + type.getName());
    }
}

File: dep/src/main/kotlin/io/vproxy/dep/vjson/util/Transformer.kt
/*
 * The MIT License
 *
 * Copyright 2021 wkgcass (https://github.com/wkgcass)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package io.vproxy.dep.vjson.util

import io.vproxy.dep.vjson.JSON
import io.vproxy.dep.vjson.simple.*
import kotlin.reflect.KClass

class Transformer {
  private val rules: MutableMap<KClass<*>, (Any) -> JSON.Instance<*>> = LinkedHashMap()

  fun <T : Any> addRule(type: KClass<T>, func: (T) -> JSON.Instance<*>): Transformer {
    @Suppress("UNCHECKED_CAST")
    rules[type] = func as (Any) -> JSON.Instance<*>
    return this
  }

  fun removeRule(type: KClass<*>): Transformer {
    rules.remove(type)
    return this
  }

  fun transform(input: Any?): JSON.Instance<*> {
    if (input == null) return SimpleNull()
    if (input is JSON.Instance<*>) {
      return input
    }
    if (input is Collection<*>) {
      val list: MutableList<JSON.Instance<*>> = ArrayList(input.size)
      for (o in input) {
        list.add(transform(o))
      }
      return object : SimpleArray(list, TrustedFlag.FLAG) {}
    }
    if (input is Map<*, *>) {
      val map: MutableList<SimpleObjectEntry<JSON.Instance<*>>> = ArrayList(input.size)
      for (key in input.keys) {
        require(key is String) { "keys of map should be String" }
        map.add(SimpleObjectEntry((key as String?)!!, transform(input[key])))
      }
      return object : SimpleObject(map, TrustedFlag.FLAG) {}
    }
    for ((key, value) in rules) {
      if (key.isInstance(input)) {
        return value(input)
      }
    }
    throw IllegalArgumentException("unknown input type: " + input::class./* #ifdef KOTLIN_JS {{ simpleName }} else {{ */qualifiedName/* }} */)
  }

  init {
    this.addRule(Boolean::class) { value -> SimpleBool(value) }
      .addRule(Int::class) { value -> SimpleInteger(value) }
      .addRule(Long::class) { value -> SimpleLong(value) }
      .addRule(Double::class) { value -> SimpleDouble(value) }
      .addRule(Float::class) { value -> SimpleDouble(value.toDouble()) }
      .addRule(Number::class) { value -> SimpleInteger(value.toInt()) }
      .addRule(String::class) { value -> SimpleString(value) }
      .addRule(Char::class) { value -> SimpleString(value.toString()) }
  }
}
